plugins/modules/win_scoop.ps1

```diff
@@ -0,0 +1,230 @@
+$global = $module.Params.global
+$name = $module.Params.name
+$no_cache = $module.Params.no_cache
+$purge = $module.Params.purge
+$skip = $module.Params.skip
+$state = $module.Params.state
+
+$module.Result.rc = 0
+
+function Install-Scoop {
+
+  # Scoop doesn't have refreshenv like Chocolatey
+  # Let's try to update PATH first
+  $env:Path = [System.Environment]::GetEnvironmentVariable("Path", "Machine") + ";" + [System.Environment]::GetEnvironmentVariable("Path", "User")
+
+  $scoop_app = Get-Command -Name scoop.ps1 -Type ExternalScript -ErrorAction SilentlyContinue
+  if ($null -eq $scoop_app) {
+    # We need to install scoop
+    # Enable TLS1.1/TLS1.2 if they're available but disabled (eg. .NET 4.5)
+    $security_protocols = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::SystemDefault
+    if ([Net.SecurityProtocolType].GetMember("Tls11").Count -gt 0) {
+      $security_protocols = $security_protcols -bor [Net.SecurityProtocolType]::Tls11
+    }
+    if ([Net.SecurityProtocolType].GetMember("Tls12").Count -gt 0) {
+      $security_protocols = $security_protcols -bor [Net.SecurityProtocolType]::Tls12
+    }
+    [Net.ServicePointManager]::SecurityProtocol = $security_protocols
+
+    $client = New-Object -TypeName System.Net.WebClient
+
+    $script_url = "https://get.scoop.sh"
+
+    try {
+      $install_script = $client.DownloadString($script_url)
+    }
+    catch {
+      $module.FailJson("Failed to download Scoop script from '$script_url'; $($_.Exception.Message)", $_)
+    }
+
+    if (-not $module.CheckMode) {
+      $res = Run-Command -Command "powershell.exe -" -stdin $install_script -environment $environment
+      if ($res.rc -ne 0) {
+        $module.Result.rc = $res.rc
+        $module.Result.stdout = $res.stdout
+        $module.Result.stderr = $res.stderr
+        $module.FailJson("Scoop bootstrap installation failed.")
+      }
+      $module.Warn("Scoop was missing from this system, so it was installed during this task run.")
+    }
+    $module.Result.changed = $true
+
+    # Refresh PATH
+    $env:Path = [System.Environment]::GetEnvironmentVariable("Path", "Machine") + ";" + [System.Environment]::GetEnvironmentVariable("Path", "User")
+
+    # locate the newly installed scoop.ps1
+    $scoop_app = Get-Command -Name scoop.ps1 -Type ExternalScript -ErrorAction SilentlyContinue
+  }
+
+  if ($module.CheckMode -and $null -eq $scoop_app) {
+    $module.Result.skipped = $true
+    $module.Result.msg = "Skipped check mode run on win_scoop as scoop.ps1 cannot be found on the system"
+    $module.ExitJson()
+  }
+
+  if ($null -eq $scoop_app -or -not (Test-Path -LiteralPath $scoop_app.Path)) {
+    $module.FailJson("Failed to find scoop.ps1, make sure it is added to the PATH")
+  }
+
+  return $scoop_app.Path
+}
+
+function Get-ScoopPackages {
+  param(
+    [Parameter(Mandatory = $true)] [string]$scoop_path
+  )
+
+  $command = Argv-ToString -arguments @("powershell.exe", $scoop_path, "export")
```

Comment:
It seems like scoop uses PowerShell scripts and is not an compiled binary. Shouldn't it be able to handle positional args as strings and we don't have to worry about escaping the arguments for the command line by just calling the script? For example this could be simplified with `$installed_packages = &$scoop_path export`.

Does scoop have a way of exporting actual objects and not strings, seems like an unfortunate omission if it isn't possible as it's bringing down the advantages that PowerShell offers.

Reply:
I gave this a go, and `scoop export` is working fine. However I am struggling with the `install` and `uninstall` commands. Here's a snipped from `Install-ScoopPackage`

```pwsh
$arguments = [System.Collections.Generic.List[String]]@("install")
$arguments.AddRange($packages)

$common_args = Get-InstallScoopPackageArguments
$arguments.AddRange($common_args)

$stdout = & $scoop_path @arguments
```

Using this I'm getting some weird errors from scoop: `Can't shim 'bin\\basename.exe': File doesn't exist.` Can you see something wrong with my PowerShell that is obviously wrong? If not, I think I'll stick with the command escaping for now.

As for PowerShell objects scoop does have this, but it's undocumented, the parameters are binary flags instead of the parameters the CLI expects, and it's recommended not to directly use the scoop internals.

Reply:
> Can you see something wrong with my PowerShell that is obviously wrong?

I can't see anything wrong with that snippet, it definitely feel like something is up with the environment PATHs.

> If not, I think I'll stick with the command escaping for now.

No worries, thought it was worth a shot to try and reduce some complexity but if it doesn't work then let's ignore it.

> As for PowerShell objects scoop does have this, but it's undocumented, the parameters are binary flags instead of the parameters the CLI expects, and it's recommended not to directly use the scoop internals.

That's a pity but it is what it is, thanks for letting me know.

Reply:
Okay. Thanks for your input. I'll resolve this comment.